subtype/src/main/java/com/fasterxml/jackson/module/subtype/SubtypeModule.java

```diff
@@ -0,0 +1,120 @@
+ * See <a href="https://github.com/FasterXML/jackson-databind/issues/2104">this issues</a> in jackson-databind.
+ * <p>
+ * When not found in the cache, it loads and caches subclasses using SPI.
+ * Therefore, we can {@link #unregisterType} a class and then module will reload this class's subclasses.
+ */
+public class SubtypeModule extends Module {
+
+    private final ConcurrentHashMap<Class<?>, List<NamedType>> subtypes = new ConcurrentHashMap<>();
+
+    @Override
+    public String getModuleName() {
+        return getClass().getSimpleName();
+    }
+
+    @Override
+    public Version version() {
+        return PackageVersion.VERSION;
+    }
+
+    @Override
+    public void setupModule(SetupContext context) {
+        context.insertAnnotationIntrospector(new AnnotationIntrospector() {
+            @Override
+            public Version version() {
+                return PackageVersion.VERSION;
+            }
+
+            @Override
+            public List<NamedType> findSubtypes(Annotated a) {
+                registerTypes(a.getRawType());
+
+                List<NamedType> list1 = SubtypeModule.findSubtypes(a.getRawType(), a::getAnnotation);
+                List<NamedType> list2 = subtypes.getOrDefault(a.getRawType(), Collections.emptyList());
+
+                if (list1.isEmpty()) return list2;
+                if (list2.isEmpty()) return list1;
+                List<NamedType> list = new ArrayList<>(list1.size() + list2.size());
+                list.addAll(list1);
+                list.addAll(list2);
+                return list;
+            }
+        });
+    }
+
+    /**
+     * load parent's subclass by SPI.
+     *
+     * @param parent parent class.
+     * @param <S>    parent class type.
+     */
+    @SuppressWarnings("unchecked")
+    public <S> void registerTypes(Class<S> parent) {
+        if (subtypes.containsKey(parent)) {
+            return;
+        }
+        List<Class<S>> subclasses = new ArrayList<>();
+        for (S instance : ServiceLoader.load(parent)) {
+            subclasses.add((Class<S>) instance.getClass());
+        }
+        this.registerTypes(parent, subclasses);
+    }
+
+    /**
+     * register subtypes without SPI.
+     * Of course, you need to provide them :)
+     *
+     * @param parent:     parent class.
+     * @param subclasses: children class.
+     * @param <S>:        parent class type.
+     */
+    public <S> void registerTypes(Class<S> parent, Iterable<Class<S>> subclasses) {
+        List<NamedType> result = new ArrayList<>();
+        for (Class<S> subclass : subclasses) {
+            result.addAll(findSubtypes(subclass, subclass::getAnnotation));
+        }
+        subtypes.put(parent, result);
+    }
+
+    public void unregisterType(Class<?> parent) {
+        subtypes.remove(parent);
+    }
+
+    private static <S> List<NamedType> findSubtypes(Class<S> clazz, Function<Class<JsonSubType>, JsonSubType> getter) {
```

Comment:
Is there a reason for keeping it static? if not maybe like...
```suggestion
    private <S> List<NamedType> _findSubtypes(Class<S> clazz, Function<Class<JsonSubType>, JsonSubType> getter) {
```